### PR TITLE
TreeDB text v2/hybrid: build comparison scoreboard

### LINKED ISSUE #2727: TreeDB text v2/hybrid: build industry comparison scoreboard
Part of #2726.

## Graph Links

- Parent: #2726
- This node: #2727

## Goal

Create the repeatable benchmark scoreboard for TreeDB text-only, vector-only, text+vector, and text+vector+scalar retrieval against relevant external baselines.

## Why This Exists

TreeDB text-v2 is now the default, and the next roadmap needs measurable progress toward industry-relevant text and hybrid retrieval performance while preserving B-tree-native storage, exact semantics, zero-doc candidate generation, and normal TreeDB maintenance paths.

## Current Evidence

Text-v2 is now default for new TreeDB collection text indexes after #2690 / PR #2707:

- Head: `6b56ec048bb851808cf72c2457b9b54c207e1d88`
- Merge: `ca321ee94cec355895bb6ffb6a59c2537c255389`
- Final artifact root: `/tmp/gomap_2690_text_v2_default_20260612_205300`

Representative current performance from that artifact:

| Shape | Current result | Notes |
| --- | ---: | --- |
| 10k text-v2 candidates, no docs | `4.23ms`, `236 ops/s` | `docs_fetched=0`, `text_state_lookups=0` |
| 10k hybrid v2 + scalar, no docs | `1.21ms`, `824 ops/s` | bounded scalar/text path |
| 100k common BM25F blockmax top-k | `3.15ms`, `318 ops/s` | `docs_fetched=0`, skipped 684 posting blocks/search |
| 100k common BM25F exhaustive top-k | `80.8ms`, `12.4 ops/s` | blockmax is ~25.7x faster |
| 100k multi-term AND, no docs | `35.5ms`, `28.1 ops/s` | exact zero-doc AND path |
| 256-doc text-v2 insert batch | `2.40ms`, `417 ops/s` | v2 ~1.93x faster than v1 row |

Correctness/contract state:

- v2 default is active for new text indexes.
- Explicit v1 opt-out remains supported and tested.
- Legacy missing-version metadata decodes as v1.
- Text/hybrid candidate generation remains zero-full-document-fetch for covered shapes.
- Text roots, posting blocks, stats, rewrite outputs, and reclamation remain B-tree-native and participate in normal TreeDB maintenance paths.

## Dependencies

None. This is the baseline dependency for later optimization tickets.

## Scope

- [ ] Define corpora, query sets, analyzer assumptions, relevance labels where available, and top-k shapes.
- [ ] Add or update harnesses for TreeDB text-v2, hybrid, vector-only, and scalar-filter shapes.
- [ ] Capture external baseline runs where practical: Lucene/Tantivy/Bleve/SQLite FTS5 for text, and Qdrant/Weaviate/Milvus/OpenSearch-style hybrid or documented local proxies where feasible.
- [ ] Report latency, throughput, allocations, storage footprint, index build time, and domain counters.

## Non-Goals And Boundaries

- No TreeDB optimization work beyond benchmark harness fixes needed for fair measurement.
- No claims of parity without reproducible external baseline commands.

## Branch, PR, Review, And Merge Policy

- Work happens on topic branches and PRs.
- Managers/subagents must not merge; coordinator owns merge after gates pass.
- Latest-head CI green, `mergeStateStatus=CLEAN`, 0 unresolved review threads, focused tests, relevant broad tests, benchmark evidence, and intended diff only are required before merge.
- Codex, Copilot, and CodeRabbit reviews should be requested only after a mature PR head: coherent code, focused tests, required benchmark evidence or rationale, current PR body/status evidence, no known local blockers, and latest-head CI running or green.
- Material regressions in runtime, throughput, p95/p99 latency, memory, allocations, storage/build/rewrite overhead, or relevant counters block merge unless optimized away or explicitly accepted with evidence.
- If CI is backed up, cancel stale non-head runs and keep latest-head CI active for the current PR head.

## Required Tests

- [ ] Harness parser tests
- [ ] Report-generation tests
- [ ] Counter validation tests for zero-doc candidate metrics

## Required Benchmarks / Performance Evidence

- [ ] 10k/100k current v2 matrix
- [ ] 1M smoke if feasible
- [ ] text-only, vector-only, hybrid, hybrid+scalar rows
- [ ] external baseline commands and artifacts

Every benchmark claim must include baseline vs candidate, exact command, hardware/context, dataset shape, measured boundary, `ns/op`, `ops/sec`, `B/op`, `allocs/op` where applicable, and relevant text/hybrid counters. Current-only snapshots are not sufficient for optimization claims.

## Completion Criteria

- [ ] Published scoreboard docs/artifacts
- [ ] Baseline commit and external versions recorded
- [ ] All later tickets can cite stable commands
- [ ] PR body includes tests, benchmarks, AI review status, unresolved thread summary, and regression assessment.
- [ ] Latest-head CI is green and review findings are fixed or explicitly resolved.

### PR DESCRIPTION
**Pull request by snissn**
## Linked issues

- Closes #2727
- Part of #2726 (M0 scoreboard foundation)

## Scope

Adds a repeatable TreeDB text-v2/hybrid industry comparison scoreboard contract:

- `scripts/bench_text_hybrid_scoreboard.sh` capture entry point.
- `cmd/treedb_text_hybrid_scoreboard` parser/report generator producing `scoreboard.json` (`treedb_text_hybrid_scoreboard/v1`) and `scoreboard.md`.
- `benchmarks/text_hybrid_scoreboard/sqlite_fts5_bench.py` local embedded SQLite FTS5 text baseline JSON artifact.
- `docs/benchmarks/treedb_text_hybrid_industry_scoreboard.md` runbook defining corpus/query/analyzer/top-k assumptions, external baseline caveats, required counters, and downstream contract surface.

Non-goals honored: no TreeDB production text/vector scoring changes, no optimizer changes, no external IR sidecar, and no industry-parity claims without artifacts/caveats.

## Start-phase plan

- Inventory existing text-v2, hybrid, vector, and external vector benchmark harnesses.
- Reuse current TreeDB benchmark rows where possible rather than adding production-path code.
- Add parser/report-generation tests and zero-doc counter validation tests.
- Capture 10k TreeDB text-only/vector-only/hybrid/hybrid+scalar rows and 100k text-v2 blockmax rows where practical.
- Capture a practical local external text baseline (SQLite FTS5) and document unavailable external systems.

## Close-phase tests

- `go test ./cmd/treedb_text_hybrid_scoreboard`
- `go test ./TreeDB/collections -run '^(TestTextV2HybridTextCandidatesNoDocCounters2627|TestTextV2DefaultSafe10KCandidateAndScalarBudgets2687|TestHybridCloseoutFixtureCandidateGenerationFetchesNoDocuments2506)$'`
- `go test ./cmd/treedb_text_hybrid_scoreboard ./TreeDB/collections`
- Smoke script check:
  - `RUN_DIR=/tmp/gomap_scoreboard_script_smoke DOCS_10K=64 BENCHTIME=1x COUNT=1 SQLITE_FTS5_DOCS=100 SQLITE_FTS5_QUERIES=5 RUN_100K=false scripts/bench_text_hybrid_scoreboard.sh`

## Benchmark evidence

Artifact root: `/tmp/gomap_2727_scoreboard_candidate_20260613_090425`

Command:

```sh
RUN_DIR=/tmp/gomap_2727_scoreboard_candidate_20260613_090425 \
BENCHTIME=1x COUNT=1 RUN_100K=true TEXT_100K_BENCHTIME=1x TEXT_100K_COUNT=1 \
SQLITE_FTS5_QUERIES=1000 \
scripts/bench_text_hybrid_scoreboard.sh
```

Hardware/context: Apple M3, 8 CPUs, macOS/Darwin arm64, Go `go1.26.0 darwin/arm64`; context file notes high background load (`load averages: 14.84 13.53 11.10`). Treat as same-host current/reference evidence, not isolated parity numbers.

Primary report: `/tmp/gomap_2727_scoreboard_candidate_20260613_090425/scoreboard.md`
JSON: `/tmp/gomap_2727_scoreboard_candidate_20260613_090425/scoreboard.json`

Representative rows from the report (all `-benchtime=1x`, `-count=1`):

| Row | Dataset/boundary | ns/op | ops/sec | B/op | allocs/op | Key counters |
| --- | --- | ---: | ---: | ---: | ---: | --- |
| TreeDB text-v2 candidates | 10k docs, no-doc `refund policy` | 5,016,250 | 199.353 | 2,969,248 | 11,181 | `docs_fetched=0`, `full_doc_fallbacks=0`, `fail_closed=0`, `text_state_lookups=0` |
| TreeDB vector candidates | 10k docs, no-doc vector | 63,125 | 15,841.6 | 149,816 | 83 | `docs_fetched=0`, `vector_candidates=64`, `vector_examined=269` |
| TreeDB hybrid | 10k docs, no-doc text+vector | 13,240,958 | 75.523 | 3,261,592 | 12,592 | `docs_fetched=0`, `text_state_lookups=0`, `vector_candidates=64` |
| TreeDB hybrid+scalar | 10k docs, no-doc text+vector+tenant filter | 2,116,333 | 472.515 | 1,999,976 | 6,291 | `docs_fetched=0`, `scalar_prefilter_ids=625`, `vector_candidates=64` |
| TreeDB text-v2 blockmax | 100k docs, no-doc common term | 3,842,292 | 260.261 | 2,557,792 | 4,105 | `docs_fetched=0`, `posting_blocks_skipped=684` |
| SQLite FTS5 | 10k docs, rowid+bm25 no-doc text baseline | 2,744,012.4 | 364.430 | n/a | n/a | `docs_fetched=0`, storage `1.28MiB`, build `0.050s` |

External baselines not run by the smoke harness are explicitly marked unavailable in the report: Lucene, Tantivy, Bleve, and Qdrant/Weaviate/Milvus/OpenSearch hybrid.

## Measurement boundary

TreeDB retrieval rows time no-document candidate generation/fusion unless explicitly labeled build/storage. Candidate-generation rows must keep `docs_fetched/search=0`, `full_doc_fallbacks/search=0`, `fail_closed/search=0`, and text-v2 state/match detail counters at zero where applicable. SQLite FTS5 times rowid + `bm25()` retrieval without primary JSON fetch.

## Performance regression assessment

This PR adds benchmark/report tooling and docs only; it does not change TreeDB production text/vector/hybrid scoring or storage paths. No material production performance regression is expected. The scoreboard artifact is current-reference evidence for downstream comparison, not an optimization claim.

## Review and CI status

- Internal high-thinking reviewer: PASS after fixes; no blocking findings/no nits.
- AI reviews: not requested yet. Per policy, they will only be requested after this mature PR is pushed with current tests/benchmark evidence and CI is running/green.
- CI: pending after PR creation.

## Unresolved threads / caveats

- 1M/10M scale runs are intentionally left to #2731; this PR provides 10k plus 100k text blockmax evidence and stable commands.
- Service-backed industry hybrid baselines require pinned external deployments before citing parity.

## Release Notes

* **New Features**
  * Added benchmarking infrastructure for SQLite FTS5 text search with detailed performance metrics (latency percentiles, throughput, result counts).
  * Added scoreboard report generator that aggregates and compares benchmark results from multiple search engines, producing formatted reports and JSON artifacts.
  * Added automated benchmark orchestration script for streamlined performance testing.

* **Documentation**
  * Added comprehensive guide for text/hybrid industry benchmarking procedures and expected metrics.

* **Tests**
  * Added test coverage for benchmarking tools and report generation.

### CODE REVIEW

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

File: cmd/treedb_text_hybrid_scoreboard/main.go [hunk 1]
```diff
@@ -0,0 +1,1272 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("parse go bench %s: %w", input.Path, err)
+	}
+	aggs := aggregateGoBenchSamples(samples)
+	rows := make([]scoreboardRow, 0, len(aggs))
+	for _, agg := range aggs {
+		row := classifyGoBenchmark(input, agg)
+		rows = append(rows, row)
+	}
+	return rows, nil
```
[the remark this answers is not in the document]
Comment (snissn) [reply cmd/treedb_text_hybrid_scoreboard/main.go:431]: Fixed in a82f8e9d0: rowsFromGoBench now rejects empty Go benchmark artifacts before report generation, and TestRowsFromGoBenchRejectsEmptyArtifacts covers a go test output with no Benchmark rows.

File: benchmarks/text_hybrid_scoreboard/sqlite_fts5_bench.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    db_path = Path(args.db) if args.db else out.with_suffix(".sqlite3")
+    if db_path.exists():
+        db_path.unlink()
+    for suffix in ("-wal", "-shm"):
+        p = Path(str(db_path) + suffix)
+        if p.exists():
+            p.unlink()
+
+    try:
+        conn = sqlite3.connect(str(db_path))
```
[the remark this answers is not in the document]
Comment (snissn) [reply benchmarks/text_hybrid_scoreboard/sqlite_fts5_bench.py:129]: Fixed in f00c5070d: sqlite_fts5_bench.py now creates both the output parent and the inferred/explicit DB parent before sqlite3.connect. Added benchmarks/text_hybrid_scoreboard/test_sqlite_fts5_bench.py to run the script with a fresh nested --out path and fail if it reports unable to open database file. Focused tests and scoreboard smoke pass.

**Comment by snissn** [comment]
## Fresh `--out` SQLite FTS5 fix (f00c5070d)

Addressed Codex P2 for direct `sqlite_fts5_bench.py --out fresh/dir/...` usage:

- Creates `out.parent` and the inferred/explicit SQLite DB parent before `sqlite3.connect`.
- Adds `benchmarks/text_hybrid_scoreboard/test_sqlite_fts5_bench.py`, which invokes the runner with a fresh nested `--out` path and fails if it reports `unable to open database file` as an unavailable baseline.
- Resolved the Codex thread with the fix summary.

Rerun evidence:

```sh
python3 -m unittest benchmarks.text_hybrid_scoreboard.test_sqlite_fts5_bench
go test ./cmd/treedb_text_hybrid_scoreboard
go test ./cmd/treedb_text_hybrid_scoreboard ./TreeDB/collections
RUN_DIR=/tmp/gomap_2727_fresh_out_fix_smoke DOCS_10K=64 BENCHTIME=1x COUNT=1 SQLITE_FTS5_DOCS=64 SQLITE_FTS5_QUERIES=5 RUN_100K=false scripts/bench_text_hybrid_scoreboard.sh
```

Fresh smoke artifact: `/tmp/gomap_2727_fresh_out_fix_smoke/scoreboard.md`.

AI review re-request status: waiting for latest-head CI on `f00c5070d` to finish, then I will re-request Codex/Copilot/CodeRabbit per the mature-PR gate.

File: scripts/bench_text_hybrid_scoreboard.sh [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+      -benchtime="$BENCHTIME" \
+      -count="$COUNT"
+
+SCOREBOARD_ARGS=(
+  -out-dir "$RUN_DIR"
+  -context "$RUN_DIR/context.txt"
+  -base-ref "origin/main"
+  -base-sha "$(git merge-base HEAD origin/main 2>/dev/null || true)"
+  -go-bench "treedb_index_insert_search_10k=$INDEX_10K"
+  -go-bench "treedb_hybrid_closeout_10k=$HYBRID_10K"
```
[the remark this answers is not in the document]
Comment (snissn) [reply scripts/bench_text_hybrid_scoreboard.sh]: Fixed in a18ae9504: the scoreboard script now uses an exact-doc hybrid source label when DOCS_10K is overridden (for example treedb_hybrid_closeout_docs_64) and writes that raw artifact name, while preserving the default treedb_hybrid_closeout_10k label for the 10k run. The report parser now recognizes exact docs_<N> source labels for benchmark outputs that lack docs_fixture/topk metrics, and TestHybridCloseoutOverrideLabelDrivesDatasetDocs plus the DOCS_10K=64 smoke artifact verify JSON/Markdown show docs=64 rather than docs=10000.

File: benchmarks/text_hybrid_scoreboard/sqlite_fts5_bench.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+def unavailable(args: argparse.Namespace, reason: str) -> dict[str, Any]:
+    return {
+        "schema_version": "treedb_text_hybrid_external/v1",
+        "status": "unavailable",
+        "system": "SQLite FTS5",
+        "engine": "sqlite_fts5",
+        "unavailable_reason": reason,
+        "command": " ".join(sys.argv),
+        "versions": versions(),
+    }
```
[the remark this answers is not in the document]
Comment (snissn) [reply benchmarks/text_hybrid_scoreboard/sqlite_fts5_bench.py]: Fixed in 23faf0027: removed the unused args parameter from unavailable() and updated the FTS5-unavailable call site.

File: benchmarks/text_hybrid_scoreboard/sqlite_fts5_bench.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+        db_path.unlink()
+    for suffix in ("-wal", "-shm"):
+        p = Path(str(db_path) + suffix)
+        if p.exists():
+            p.unlink()
+
+    try:
+        conn = sqlite3.connect(str(db_path))
+        configure(conn)
+        conn.execute(f"CREATE VIRTUAL TABLE docs_fts USING fts5(title, body, tokenize='{args.tokenize}')")
```
[the remark this answers is not in the document]
Comment (snissn) [reply benchmarks/text_hybrid_scoreboard/sqlite_fts5_bench.py]: Fixed in 23faf0027: --tokenize now must match a strict allowlist (unicode61/ascii/porter) before any DB open/DDL, and the CREATE VIRTUAL TABLE SQL is selected from constant allowlisted DDL strings instead of interpolating raw argv. Added test_rejects_unsupported_tokenizer_before_sql_ddl to cover an injection-shaped value.

File: benchmarks/text_hybrid_scoreboard/sqlite_fts5_bench.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+        raise SystemExit(f"query {args.query!r} returned no rows")
+
+    durations: list[int] = []
+    total_results = 0
+    search_start = time.perf_counter_ns()
+    for i in range(args.queries):
+        start = time.perf_counter_ns()
+        rows = list(conn.execute(sql, (args.query, args.top_k)))
+        durations.append(time.perf_counter_ns() - start)
+        total_results += len(rows)
```
[the remark this answers is not in the document]
Comment (snissn) [reply benchmarks/text_hybrid_scoreboard/sqlite_fts5_bench.py]: Fixed in 23faf0027: renamed the unused benchmark-loop variable to _.

File: benchmarks/text_hybrid_scoreboard/test_sqlite_fts5_bench.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+                    "--top-k",
+                    "5",
+                    "--out",
+                    str(out),
+                ],
+                check=False,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                text=True,
+            )
```
[the remark this answers is not in the document]
Comment (snissn) [reply benchmarks/text_hybrid_scoreboard/test_sqlite_fts5_bench.py:35]: Fixed in 23faf0027: replaced explicit stdout/stderr PIPEs with capture_output=True in the focused Python runner test.

File: cmd/treedb_text_hybrid_scoreboard/main_test.go [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+	if got, want := len(rep.CounterValidations), 2; got != want {
+		t.Fatalf("counter validations=%d want %d", got, want)
+	}
+	md := renderMarkdown(rep)
+	for _, want := range []string{"SQLite FTS5", "text_scalar", "Lucene", "not run in unit test", "scoreboard"} {
+		if !strings.Contains(md, want) {
+			t.Fatalf("markdown missing %q:\n%s", want, md)
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (snissn) [reply cmd/treedb_text_hybrid_scoreboard/main_test.go:202]: Reviewed. I am intentionally keeping this as a cohesive integration-style report test: it exercises buildReport input parsing, counter validation, unavailable rows, and Markdown rendering over one shared fixture. Splitting would duplicate setup without increasing coverage for this PR, and CodeRabbit marked the tradeoff as optional/also reasonable. No code change.

File: cmd/treedb_text_hybrid_scoreboard/main_test.go [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+func TestRunRejectsMissingInputsEvenWhenCounterFailuresAllowed(t *testing.T) {
+	dir := t.TempDir()
+	err := run(config{outDir: filepath.Join(dir, "out"), allowCounterFailures: true, goBenches: namedPaths{{Name: "missing", Path: filepath.Join(dir, "missing.txt")}}})
```
[the remark this answers is not in the document]
Comment (snissn) [reply cmd/treedb_text_hybrid_scoreboard/main_test.go]: Fixed in 23faf0027: reformatted the long config literal in TestRunRejectsMissingInputsEvenWhenCounterFailuresAllowed.

**Comment by snissn** [comment]
## CodeRabbit follow-up fixes (23faf0027)

Addressed/answered all 6 latest-head CodeRabbit threads:

- `sqlite_fts5_bench.py`: validates `--tokenize` against an allowlist (`unicode61`, `ascii`, `porter`) before DB open/DDL and executes only constant allowlisted FTS5 DDL strings.
- Added `test_rejects_unsupported_tokenizer_before_sql_ddl` for an injection-shaped tokenizer value.
- Removed unused `unavailable(args, ...)` parameter and updated call site.
- Renamed unused benchmark loop variable to `_`.
- Switched focused Python subprocess test to `capture_output=True`.
- Reformatted the long Go test config literal.
- Kept the cohesive report-builder integration test intentionally; replied in-thread with rationale because it validates the shared buildReport/counter/Markdown fixture end-to-end.

Rerun evidence:

```sh
python3 -m unittest benchmarks.text_hybrid_scoreboard.test_sqlite_fts5_bench
go test ./cmd/treedb_text_hybrid_scoreboard -count=1
go test ./TreeDB/collections -count=1 -run '^(TestTextV2HybridTextCandidatesNoDocCounters2627|TestTextV2DefaultSafe10KCandidateAndScalarBudgets2687|TestHybridCloseoutFixtureCandidateGenerationFetchesNoDocuments2506)$'
go test ./cmd/treedb_text_hybrid_scoreboard ./TreeDB/collections
RUN_DIR=/tmp/gomap_2727_coderabbit_fix_smoke DOCS_10K=64 BENCHTIME=1x COUNT=1 SQLITE_FTS5_DOCS=64 SQLITE_FTS5_QUERIES=5 RUN_100K=false scripts/bench_text_hybrid_scoreboard.sh
```

Smoke artifact: `/tmp/gomap_2727_coderabbit_fix_smoke/scoreboard.md`; JSON hybrid closeout rows use `treedb_hybrid_closeout_docs_64` and `docs=64` (no `docs=10000` hybrid rows).

Latest-head CI for `23faf0027` is running; prior head `a18ae9504` was fully green.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review

File: cmd/treedb_text_hybrid_scoreboard/main.go [hunk 10]
```diff
@@ -0,0 +1,1286 @@
...
+		Context:       ctx,
+		Commands:      []namedValue(cfg.commands),
+		Inputs: reportInputs{
+			GoBenchmarks: []namedPath(cfg.goBenches),
+			ExternalJSON: []namedPath(cfg.externals),
+		},
+		Caveats: append([]string(nil), cfg.caveats...),
+	}
+	if len(rep.Caveats) == 0 {
+		rep.Caveats = defaultCaveats()
```
[the remark this answers is not in the document]
Comment (snissn) [reply cmd/treedb_text_hybrid_scoreboard/main.go]: Fixed in 8b0b9d4c: buildReport now initializes Caveats with defaultCaveats() and appends any custom -caveat values, so the default script RUN_100K=false caveat no longer drops same-host/zero-doc warnings. Added TestBuildReportAppendsCustomCaveatsToDefaults, and the DOCS_10K=64 smoke artifact verifies scoreboard.json/Markdown include both default caveats and the custom 100k smoke caveat.

**Comment by snissn** [comment]
## Codex caveat follow-up (8b0b9d4c)

Addressed latest Codex P2 on custom `-caveat` handling:

- `buildReport` now always starts with `defaultCaveats()` and appends any custom `-caveat` entries.
- The default script path (`RUN_100K=false`) therefore keeps built-in same-host/local-benchmark and zero-doc counter warnings while also adding the 100k opt-in caveat.
- Added `TestBuildReportAppendsCustomCaveatsToDefaults`.
- Resolved the Codex thread with the fix summary.

Rerun evidence:

```sh
go test ./cmd/treedb_text_hybrid_scoreboard -count=1
go test ./TreeDB/collections -count=1 -run '^(TestTextV2HybridTextCandidatesNoDocCounters2627|TestTextV2DefaultSafe10KCandidateAndScalarBudgets2687|TestHybridCloseoutFixtureCandidateGenerationFetchesNoDocuments2506)$'
go test ./cmd/treedb_text_hybrid_scoreboard ./TreeDB/collections
RUN_DIR=/tmp/gomap_2727_caveat_append_smoke DOCS_10K=64 BENCHTIME=1x COUNT=1 SQLITE_FTS5_DOCS=64 SQLITE_FTS5_QUERIES=5 RUN_100K=false scripts/bench_text_hybrid_scoreboard.sh
```

Smoke artifact: `/tmp/gomap_2727_caveat_append_smoke/scoreboard.md`; caveats include both defaults and the 100k opt-in caveat.

Latest-head CI is running for `8b0b9d4c`.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review

File: cmd/treedb_text_hybrid_scoreboard/main.go [hunk 11]
```diff
@@ -0,0 +1,1283 @@
...
+	if commit == "" {
+		commit = strings.TrimSpace(runGit(repoRoot, "rev-parse", "HEAD"))
+	}
+	contextText := ""
+	if cfg.contextPath != "" {
+		if raw, err := os.ReadFile(cfg.contextPath); err == nil {
+			contextText = strings.TrimSpace(string(raw))
+		}
+	}
+	goVersion := strings.TrimSpace(runCmd("go", "version"))
```
[the remark this answers is not in the document]
Comment (snissn) [reply cmd/treedb_text_hybrid_scoreboard/main.go]: Fixed in 696a6c4fa: scoreboard context now prefers the captured context.txt go= line written by scripts/bench_text_hybrid_scoreboard.sh via GO_BIN version, and falls back to the generator runtime version instead of shelling out to plain go from PATH. Added TestBuildReportUsesCapturedGoVersionFromContext and verified with a fake GO_BIN smoke run that scoreboard.json/Markdown record go version go9.99.0 fake/scoreboard.

**Comment by snissn** [comment]
## Codex Go toolchain metadata follow-up (696a6c4fa)

Addressed latest Codex P2 on Go version metadata:

- `buildContext` now prefers the captured `context.txt` `go=` line written by `scripts/bench_text_hybrid_scoreboard.sh` using `$GO_BIN version`.
- If no captured context is available, the report falls back to the generator runtime version (`runtime.Version()` + `GOOS/GOARCH`) instead of shelling out to plain `go` from `PATH`.
- Added `TestBuildReportUsesCapturedGoVersionFromContext`.
- Resolved the Codex thread with the fix summary.

Rerun evidence:

```sh
go test ./cmd/treedb_text_hybrid_scoreboard -count=1
go test ./TreeDB/collections -count=1 -run '^(TestTextV2HybridTextCandidatesNoDocCounters2627|TestTextV2DefaultSafe10KCandidateAndScalarBudgets2687|TestHybridCloseoutFixtureCandidateGenerationFetchesNoDocuments2506)$'
go test ./cmd/treedb_text_hybrid_scoreboard ./TreeDB/collections
python3 -m unittest benchmarks.text_hybrid_scoreboard.test_sqlite_fts5_bench
RUN_DIR=/tmp/gomap_2727_go_version_smoke GO_BIN=/tmp/gomap_2727_fake_go DOCS_10K=64 BENCHTIME=1x COUNT=1 SQLITE_FTS5_DOCS=64 SQLITE_FTS5_QUERIES=5 RUN_100K=false scripts/bench_text_hybrid_scoreboard.sh
```

Smoke artifact: `/tmp/gomap_2727_go_version_smoke/scoreboard.md`; both JSON and Markdown record `go version go9.99.0 fake/scoreboard` from the fake `$GO_BIN version` capture.

Latest-head CI is running for `696a6c4fa`.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review

**Comment by snissn** [comment]
## Mergeability handoff for #2727 / #2735 (696a6c4fa)

Status: mergeable candidate; not merged.

Latest blocker fix:
- Scoreboard metadata now uses the captured `context.txt` `go=` line from the benchmark script instead of shelling out to plain `go version`, so report metadata matches `GO_BIN` used for capture.
- Added focused coverage for captured Go version metadata.

Validation:
- `go test ./cmd/treedb_text_hybrid_scoreboard -count=1`
- focused TreeDB hybrid/counter tests
- `go test ./cmd/treedb_text_hybrid_scoreboard ./TreeDB/collections`
- `python3 -m unittest benchmarks.text_hybrid_scoreboard.test_sqlite_fts5_bench`
- `git diff --check`
- Fake-`GO_BIN` smoke: `/tmp/gomap_2727_go_toolchain_metadata_smoke_latest/scoreboard.json` records commit `696a6c4fa23dfe0b0f703a0af1aca0399073834d` and `go version go9.99.0 fake/scoreboard`.

CI/reviews:
- All PR checks currently pass (26/26), including `race-check (linux)-2` after rerun. The prior failure was an artifact upload `ECONNRESET` after tests passed; rerun passed.
- Review threads: 11 total, 0 unresolved.
- Re-requested Codex, Copilot, and CodeRabbit reviews after the latest push; Codex reported no major issues and CodeRabbit status is passing.

**Comment by snissn** [comment]
## Coordinator merge approval

Verified latest head `696a6c4fa23dfe0b0f703a0af1aca0399073834d` for #2727.

Gates:
- `mergeStateStatus=CLEAN`, branch not draft.
- Latest-head CI: 26/26 checks passing. Earlier `race-check (linux)-2` failure was artifact-upload `ECONNRESET` after race tests passed; rerun passed.
- Review threads: 0 unresolved.
- AI review: Codex latest-head review reports no major issues; CodeRabbit passing; Copilot requested.
- Scope review: benchmark/report/docs only; no TreeDB production scoring/storage path changes.

Coordinator spot-checks:
- `git diff --check origin/main...HEAD`
- `go test ./cmd/treedb_text_hybrid_scoreboard`
- `python3 -m unittest benchmarks.text_hybrid_scoreboard.test_sqlite_fts5_bench`
- `go test ./cmd/treedb_text_hybrid_scoreboard ./TreeDB/collections -run '^(TestTextV2HybridTextCandidatesNoDocCounters2627|TestTextV2DefaultSafe10KCandidateAndScalarBudgets2687|TestHybridCloseoutFixtureCandidateGenerationFetchesNoDocuments2506)$'`
- Smoke: `RUN_DIR=/tmp/gomap_2727_coordinator_smoke DOCS_10K=32 BENCHTIME=1x COUNT=1 SQLITE_FTS5_DOCS=32 SQLITE_FTS5_QUERIES=2 RUN_100K=false scripts/bench_text_hybrid_scoreboard.sh`

Merging by coordinator per roadmap policy.
